katdal/datasources.py

```diff
@@ -228,34 +240,78 @@ def __init__(self, store, chunk_info, corrprods):
+                src_keys[index] = (array.name,) + index
             # array may have fewer dimensions than flags
             # (specifically, for weights_channel).
-            if has_array.ndim < darray['flags'].ndim:
-                chunks += tuple((x,) for x in darray['flags'].shape[has_array.ndim:])
+            chunks = array.chunks
+            if array.ndim < darray['flags'].ndim:
+                chunks += tuple((x,) for x in darray['flags'].shape[array.ndim:])
             intersections = intersect_chunks(darray['flags'].chunks, chunks)
-            for has, pieces in zip(has_array.flat, intersections):
-                if not has:
-                    for piece in pieces:
-                        chunk_idx, slices = zip(*piece)
-                        lost[chunk_idx].append(slices)
-        # 'literal' is needed to prevent dask copying the dictionary for every
-        # chunk.
-        flags = da.map_blocks(_apply_data_lost, darray['flags'], dtype=np.uint8,
-                              name=flags_raw_name, lost=literal(lost))
+            src_indices = itertools.product(*(range(len(c)) for c in array.chunks))
+            for src_key, pieces in zip(src_keys.flat, intersections):
+                for piece in pieces:
+                    dst_index, slices = zip(*piece)
+                    # if src_index is missing, then the parts of dst_index
+                    # indicated by slices must be flagged.
+                    # TODO: fast path for when slices covers the whole chunk?
+                    lost_map[dst_index].extend([src_key, slices])
+        dsk = {
+            (flags_raw_name,) + key: (
+                _apply_data_lost,
+                (flags_orig_name,) + key,
+                value
+            ) for key, value in np.ndenumerate(lost_map)
+        }
+        dsk = HighLevelGraph.from_collections(
+            flags_raw_name, dsk, dependencies=list(darray.values())
+        )
+        flags = da.Array(dsk, flags_raw_name,
+                         chunks=darray['flags'].chunks,
+                         shape=darray['flags'].shape,
+                         dtype=darray['flags'].dtype)
+
+        # Turn missing blocks in the other arrays into zeros to make them
+        # valid dask arrays.
+        for array_name, array in darray.items():
+            if array_name == 'flags':
+                continue
+            new_name = 'filled-' + array.name
+            indices = itertools.product(*(range(len(c)) for c in array.chunks))
+            dsk = {
+                (new_name,) + index: (
+                    _default_zero,
+                    (array.name,) + index,
+                    shape,
+                    array.dtype
+                ) for index, shape in zip(indices, itertools.product(*array.chunks))
+            }
+            dsk = HighLevelGraph.from_collections(new_name, dsk, dependencies=[array])
+            darray[array_name] = da.Array(dsk, new_name,
```

Comment:
So the "filled" dask arrays replace the ones in `darray` but the new `flags` array doesn't?

Reply:
Good question. It doesn't actually matter because it doesn't get used again, but I've updated the code to update `darray['flags']` with the fixed-up flags.